src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1305 @@
+ * executor. This backend attaches to a shared memory segment as identified by the
+ * main_arg of the background worker.
+ *
+ * This is mostly based on the background worker logic in pg_cron
+ */
+void
+CitusBackgroundTaskExecuter(Datum main_arg)
+{
+	/*
+	 * TODO figure out if we need this signal handler that is in pgcron
+	 * pqsignal(SIGTERM, pg_cron_background_worker_sigterm);
+	 */
+	BackgroundWorkerUnblockSignals();
+
+	/* Set up a memory context and resource owner. */
+	Assert(CurrentResourceOwner == NULL);
+	CurrentResourceOwner = ResourceOwnerCreate(NULL, "citus background job");
+	CurrentMemoryContext = AllocSetContextCreate(TopMemoryContext,
+												 "citus background job execution",
+												 ALLOCSET_DEFAULT_MINSIZE,
+												 ALLOCSET_DEFAULT_INITSIZE,
+												 ALLOCSET_DEFAULT_MAXSIZE);
+
+	/* Set up a dynamic shared memory segment. */
+	dsm_segment *seg = dsm_attach(DatumGetInt32(main_arg));
+	if (seg == NULL)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 errmsg("unable to map dynamic shared memory segment")));
+	}
+
+	shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC, dsm_segment_address(seg));
+	if (toc == NULL)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 errmsg("bad magic number in dynamic shared memory segment")));
+	}
+
+	char *database = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_DATABASE, false);
+	char *username = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_USERNAME, false);
+	char *command = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_COMMAND, false);
+	int64 *taskId = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_TASK_ID, false);
+	shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+
+	shm_mq_set_sender(mq, MyProc);
+	shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+	pq_redirect_to_shm_mq(seg, responseq);
+
+	/* setup error context to indicate the errors came from a running background task */
+	ErrorContextCallback errorCallback = { 0 };
+	CitusBackgroundJobExecuterErrorCallbackContext context = {
+		.database = database,
+		.username = username,
+	};
+	errorCallback.callback = CitusBackgroundJobExecuterErrorCallback;
+	errorCallback.arg = (void *) &context;
+	errorCallback.previous = error_context_stack;
+	error_context_stack = &errorCallback;
+
+	BackgroundWorkerInitializeConnection(database, username, 0);
+
+	/* make sure we are the only backend running for this task */
+	LOCKTAG locktag = { 0 };
+	SET_LOCKTAG_BACKGROUND_TASK(locktag, *taskId);
+	const bool sessionLock = true;
+	const bool dontWait = true;
+	LockAcquireResult locked =
+		LockAcquire(&locktag, AccessExclusiveLock, sessionLock, dontWait);
+	if (locked == LOCKACQUIRE_NOT_AVAIL)
+	{
+		ereport(ERROR, (errmsg("unable to acquire background task lock for taskId: %ld",
+							   *taskId),
+						errdetail("this indicates that an other backend is already "
+								  "executing this task")));
+		exit(0);
```

Comment:
```suggestion
```
minor: this exit(0) is a bit confusing since the actual exit code is 1 and seems unlikely to be necessary for any compiler given that flow continues below